lmnet/lmnet/datasets/tensorflow_datasets.py

```diff
@@ -0,0 +1,141 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+import numpy as np
+import tensorflow as tf
+import tensorflow_datasets as tfds
+
+from lmnet.datasets.base import Base, ObjectDetectionBase
+from lmnet import data_processor
+
+
+class TensorFlowDatasetsBase(Base):
+    """
+    Abstract dataset class for loading TensorFlow Datasets.
+    The parameter "extend_dir" should be a path to the prepared data of tfds.
+    "extend_dir" can be a path either in local or in Google Cloud Storage.
+    """
+    available_subsets = ["train", "train_validation_saving", "validation"]
+
+    def __init__(
+            self,
+            *args,
+            **kwargs
+    ):
+        super().__init__(
+            *args,
+            **kwargs,
+        )
+
+        # The last directory name of extend_dir is the dataset name
+        splitted_path = self.extend_dir.rstrip('/').split('/')
+        data_dir = '/'.join(splitted_path[:-1])
+        name = splitted_path[-1]
+        builder = tfds.builder(name, data_dir=data_dir)
+
+        self.info = builder.info
+        self._init_available_splits()
+
+        self.dataset = builder.as_dataset(split=self.available_splits[self.subset])
+        self._init_features()
+
+    def _init_available_splits(self):
+        """
+        Initializing available splits dictionary depending on
+        what kind of splits the dataset has.
+        """
+        self.available_splits = {}
+        if tfds.Split.TRAIN not in self.info.splits:
+            raise ValueError("Datasets need to have a split \"TRAIN\".")
+
+        if tfds.Split.VALIDATION in self.info.splits and tfds.Split.TEST in self.info.splits:
+            self.available_splits["train"] = tfds.Split.TRAIN
+            self.available_splits["validation"] = tfds.Split.VALIDATION
+            self.available_splits["train_validation_saving"] = tfds.Split.TEST
+
+        elif tfds.Split.VALIDATION in self.info.splits:
+            self.available_splits["train"] = tfds.Split.TRAIN
+            self.available_splits["validation"] = tfds.Split.VALIDATION
+
+        elif tfds.Split.TEST in self.info.splits:
+            self.available_splits["train"] = tfds.Split.TRAIN
+            self.available_splits["validation"] = tfds.Split.TEST
+
+        else:
+            train, validation = tfds.Split.TRAIN.subsplit([9, 1])
+            self.available_splits["train"] = train
+            self.available_splits["validation"] = validation
+
+    def _init_features(self):
+        """
+        Initializing feature variables by using tf.Tensor from datasets.
+        For classification, "images" and "labels" should be initialized.
+        For object detection, "images" and "gt_boxes" should be initialized.
+        """
+        raise NotImplementedError()
+
+    @property
+    def num_per_epoch(self):
+        return len(self.images)
```

Comment:
Maybe `builder.info` has the information of data number.

Reply:
The number of data in pre-defined split (e.g. TRAIN, VALIDATION, TEST) is included in `builder.info`. But TRAIN split will be splitted into two sub-splits if datasets only have TRAIN split.
https://github.com/tfujiwar/blueoil/blob/828d152d7f5d04cda8be9000ff9b46e7847b3d4b/lmnet/lmnet/datasets/tensorflow_datasets.py#L77-L80

That's why `len(self.images)` is used here.